bullet-featherstone/src/Base.cc

```diff
@@ -26,14 +26,213 @@ namespace gz {
+  int numManifolds = this->getNumManifolds();
+  for (int i = 0; i < numManifolds; ++i)
+  {
+    btPersistentManifold* contactManifold =
+        this->getManifoldByIndexInternal(i);
+
+    const btMultiBodyLinkCollider* ob0 =
+        dynamic_cast<const btMultiBodyLinkCollider *>(
+        contactManifold->getBody0());
+    const btMultiBodyLinkCollider* ob1 =
+        dynamic_cast<const btMultiBodyLinkCollider *>(
+        contactManifold->getBody1());
+
+    // if it's a custom manifold, just refresh contacts, no need to
+    // loop through and check them.
+    if (this->customManifolds.find(contactManifold) !=
+        this->customManifolds.end())
+    {
+      contactManifold->refreshContactPoints(ob0->getWorldTransform(),
+                                            ob1->getWorldTransform());
+      continue;
+    }
+
+    if (this->manifoldsToKeep.find(contactManifold) !=
+        this->manifoldsToKeep.end())
+    {
+      continue;
+    }
+
+    const btCompoundShape *compoundShape0 =
+        static_cast<const btCompoundShape *>(ob0->getCollisionShape());
+    const btCompoundShape *compoundShape1 =
+        static_cast<const btCompoundShape *>(ob1->getCollisionShape());
+
+    int numContacts = contactManifold->getNumContacts();
+    for (int j = 0; j < numContacts; ++j)
+    {
+      btManifoldPoint& pt = contactManifold->getContactPoint(j);
+      const btCollisionShape *colShape0 = this->FindCollisionShape(
+          compoundShape0, pt.m_index0);
+      const btCollisionShape *colShape1 = this->FindCollisionShape(
+          compoundShape1, pt.m_index1);
+
+      if (!colShape0 || !colShape1 ||
+         (!this->HasConvexHullChildShapes(colShape0) &&
+          !this->HasConvexHullChildShapes(colShape1)))
+      {
+        this->manifoldsToKeep.insert(contactManifold);
+        continue;
+      }
+
+      btPersistentManifold* colManifold =
+          this->colPairManifolds[colShape0][colShape1];
+      if (!colManifold)
+      {
+        // create new custom manifold for the collision pair
+        colManifold = this->getNewManifold(ob0, ob1);
+        this->colPairManifolds[colShape0][colShape1] = colManifold;
+        this->colPairManifolds[colShape1][colShape0] = colManifold;
+        this->customManifolds.insert(colManifold);
+      }
+      colManifold->addManifoldPoint(pt);
+      colManifold->refreshContactPoints(ob0->getWorldTransform(),
+                                        ob1->getWorldTransform());
+    }
+
+    // clear manifolds that are replaced by custom ones
+    if (this->manifoldsToKeep.find(contactManifold) ==
+        this->manifoldsToKeep.end())
+      contactManifold->clearManifold();
+  }
+}
+
+/////////////////////////////////////////////////
+void GzCollisionDispatcher::releaseManifold(btPersistentManifold *_manifold)
+{
+  auto manifoldIt = this->manifoldsToKeep.find(_manifold);
+  if (manifoldIt != this->manifoldsToKeep.end())
+    this->manifoldsToKeep.erase(manifoldIt);
+
```

Comment:
Do we need to also remove it from `customManifolds`?

Reply:
I think this is only be called for contact manifolds owned by bullet, and not the custom ones created by gz-physics. I tested manually by removing models with convex decomposed collisions and did not notice that custom manifolds were passed to this function.